### PR TITLE
Enh verbose and no execute options

### PR DESCRIPTION
**Pull request by hardikpnsp**
### Enhancement

- Added option `-v` for verbose where all curl and tar commands that are executed are echoed.
- Added option `-n` for no-execute where nothing is executed but if `-v` is passed it will still echo the commands to be executed. 
- Added new test cases for the above options. 

### Checks

- [x] all tests pass 
- [x] docs updated

Closes issue: (#10)

### CODE REVIEW

**Comment by olavfosse** [comment]
Thanks for the PR.

I think it's best if we put the flags right before or after the subcommands(file/url/shorten) and prior to the file/url.

I will push a commit that updates the manpage to describe the desired usage.

**Comment by olavfosse** [comment]
It is probably worth looking into **GETOPT(1)** for parsing the options. For the program to be predictable it is important that it reads arguments in a normal way. It should allow all sorts of variations like `-v -n`, `-n -v`, `-vn`, `-nv` and also odd ones like `-nvnvnnnnv -nvnvn -n -v -nvn`. I think this should be doable if you use **GETOPTS(1)**.

**Comment by hardikpnsp** [comment]
used getopt for getting the options
now it works as:
```bash
0x0 file [OPTIONS] filename  
```
there were conflicts in the pdf file, as it's a binary file I ended up using the one in the current branch.

**Comment by olavfosse** [comment]
> there were conflicts in the pdf file, as it's a binary file I ended up using the one in the current branch.

Yeah, the way to resolve that is just to resolve doc/0x0.1 and generate a new pdf with bin/doc.sh. That script uses mandoc though which you might be missing. Don't worry about it, I'll handled the merge conflicts :).

**Comment by olavfosse** [comment]
Why is it that the curl commands have a closing parentheses appended to them?

**Comment by hardikpnsp** [comment]
> Why is it that the curl commands have a closing parentheses appended to them?

That seems like a typo, let me check.

**Comment by olavfosse** [comment]
To DRY the code we should use a function like `maybe_run_and_maybe_print`, or something like that, instead of WET.

File: src/0x0 [hunk 1]
```diff
@@ -80,17 +83,22 @@ request() {
...
-		4** | 5**)
-			exit_with_error "error: $(get_full_status_code "$code")"
-			;;
-		*)
-			echo "$response"
-	esac
+	if $verbose ; then
+		echo "\$ curl -Ss -w 'status_code=%{http_code}' https://0x0.st "$form_string")"
+	fi
+	if ! $no_execute ; then
```
Comment (olavfosse) [inline src/0x0]: WET

File: src/0x0 [hunk 2]
```diff
@@ -104,7 +112,14 @@ dispatch_file() {
 	elif [ -d "$file" ]; then
 		# stderr is redirected to /dev/null to prevent the following from being displayed
 		# tar: Removing leading / from absolute path names in the archive
-		tar cf - "$file" 2> /dev/null | request file - || exit_with_error 'error: tar archival failed'
+		if $verbose ; then
+			echo "\$ tar cf - "$file" 2> /dev/null"
+		fi
+		if ! $no_execute ; then
```
Comment (olavfosse) [inline src/0x0]: WET

**Comment by olavfosse** [comment]
When I think about it, using a function like `maybe_run_and_maybe_print` is not as straight forward as I initially thought. This is because the output of the curl command is further processed and separating the command (to be displayed to the user if -v) and the output of the command sounds quite hackish. We can just assign the command to a variable and print/run that instead.

File: src/0x0 [hunk 3]
```diff
@@ -77,57 +77,73 @@ form_string() {
...
 		# tar: Removing leading / from absolute path names in the archive
-		tar cf - "$file" 2> /dev/null | request file - || exit_with_error 'error: tar archival failed'
+		if "$verbose"; then
+			echo "\$ tar cf - "$file" 2> /dev/null"
+		fi
+		if "$execute"; then
+			tar cf - "$file" 2> /dev/null | request file - "$execute" "$verbose" || exit_with_error 'error: tar archival failed'
+		else
+			request file - "$execute" "$verbose"
+		fi
```
Comment (olavfosse) [inline src/0x0]: ?

Comment (olavfosse) [reply src/0x0]: My bad, we need to run it in case verbose is enabled.

Comment (olavfosse) [reply src/0x0]: I thought, it was redundant at first. I added a comment that says it's not.

**Comment by olavfosse** [comment]
Tests are passing on MacOS, OpenBSD and Ubuntu. Merging.

**Pull request merged**